### PR TITLE
release v4.5.0

### PR DESCRIPTION
**Pull request by jameslamb**
### Release checklist:

Copied from #6439 with a few changes.

**deadline**

August 3, 2024 (ref: #6522)

**before merge**

- [x] Update [VERSION.txt](https://github.com/microsoft/LightGBM/blob/master/VERSION.txt) number.
- [x] Update version in [Appveyor config file](https://github.com/microsoft/LightGBM/blob/master/.appveyor.yml).
- [x] Update version in [`configure` file of R-package](https://github.com/microsoft/LightGBM/blob/master/R-package/configure): `/gha run r-configure`.
- [x] Change `development.mode` from `unreleased` to `release` in [pkgdown config file](https://github.com/microsoft/LightGBM/blob/master/R-package/pkgdown/_pkgdown.yml).
- [x] Update version in `python-package/pyproject.toml`
- [x] Add release branch to [RTD versions](https://readthedocs.org/projects/lightgbm/versions/), trigger a new build, check docs
- [x] All docs for new behavior have Sphinx `versionadded::` annotations ([docs on those](https://www.sphinx-doc.org/en/master/usage/restructuredtext/directives.html#directive-versionadded))
- [x] All new parameters in `config.h` have `*New in version {version}*` comments added
- [x] Run the valgrind checks with `/gha run r-valgrind` ([docs link](https://github.com/microsoft/LightGBM/blob/master/R-package/README.md#valgrind)) 
- [x] ~test Python and R packages on arm64 Mac~
  - *This now always happens in CI, thanks to #6391 and #6494. Let's remove it from this checklist in the next release.*

Items to be addressed before merging:

* [x] #6524
* [x] #6489
* [x] #6523

**after merge**

- [ ] Push `v*` tag to trigger `GitHubRelease` action at Azure Pipelines.
- [ ] Update [`stable` tag at GitHub](https://github.com/microsoft/LightGBM/releases/tag/stable).

<details><summary>code to update tags (click me)</summary>

```shell
NEW_VERSION="4.4.0"
git checkout master
git pull upstream master
git fetch upstream --tags
git tag -d stable
git push upstream :refs/tags/stable
git tag stable
git tag "v${NEW_VERSION}"
git push upstream stable "v${NEW_VERSION}"
```

</details>

- [ ] Add new tag to [RTD versions](https://readthedocs.org/projects/lightgbm/versions/) and trigger a new build.
- [ ] Remove the release branch `release/v{n}.{n}.{n}` from RTD versions

**after the release is created by Azure DevOps job**

- [ ] Add release notes from `release-drafter` to the automatically-created GitHub release, convert it from `Draft` to published.
- [ ] Upload release to [CRAN](https://cran.r-project.org/package=lightgbm)
- [ ] Upload release to [test PyPI](https://test.pypi.org/project/lightgbm/)
- [ ] Upload release to [PyPI](https://pypi.org/project/lightgbm/).

<details><summary>code to publish to PyPI (summary)</summary>

```shell
mkdir /tmp/lgb-release
cd /tmp/lgb-release

NEW_VERSION="4.4.0"

# NOTE: requires a GitHub personal access token with "repo" scope
gh release download \
    --repo microsoft/LightGBM \
    --dir ./artifacts \
    --pattern 'lightgbm*-py3-*.whl' \
    --pattern "lightgbm-${NEW_VERSION}.tar.gz" \
    "v${NEW_VERSION}"

# NOTE: manually downloaded arm64 wheels from CI

# config docs: https://packaging.python.org/en/latest/specifications/pypirc/
twine upload \
    -r testpypi \
    ./artifacts/*
```

*(`gh` is the GitHub CLI, see https://cli.github.com/manual/gh_release_download)*

Then confirmed that installing the latest wheel works.

```shell
pip install -i https://test.pypi.org/simple/ "lightgbm==${NEW_VERSION}"
python ./examples/python-guide/logistic_regression.py
```

Then pushed them to real PyPI.

```shell
twine upload \
    ./artifacts/*
```

</details>

- [ ] Update [cran-comments.md](https://github.com/microsoft/LightGBM/blob/master/R-package/cran-comments.md) when new release is accepted on CRAN.
- [ ] Upload release to [NuGet](https://www.nuget.org/packages/LightGBM/).
- [ ] Update version and commit hash in [Homebrew formula](https://github.com/Homebrew/homebrew-core/blob/master/Formula/l/lightgbm.rb) ([docs](https://github.com/Homebrew/homebrew-core/blob/HEAD/CONTRIBUTING.md#to-submit-a-version-upgrade-for-the-foo-formula))
- [ ] put up another PR incrementing development version here in GitHub (example: #6090)

<details><summary>code to open that dev version PR (click me)</summary>

```shell
RELEASE_PR_NUMBER="6439"
RELEASE_VERSION=$(cat ./VERSION.txt)
DEV_VERSION="${RELEASE_VERSION}.99"

git checkout -b ci/dev-version

echo "${DEV_VERSION}" > ./VERSION.txt
sed \
    -i .bak \
    "s|${RELEASE_VERSION}|${DEV_VERSION}|g" \
    .appveyor.yml

sed \
    -i .bak \
    "s|version = \"${RELEASE_VERSION}\"|version = \"${DEV_VERSION}\"|g" \
    ./python-package/pyproject.toml

sed \
    -i .bak \
    's|mode\: release|mode\: unreleased|g' \
    ./R-package/pkgdown/_pkgdown.yml

docker run \
    --rm \
    -v $(pwd):/opt/LightGBM \
    -w /opt/LightGBM \
    ubuntu:22.04 \
    ./R-package/recreate-configure.sh

git add \
    ./.appveyor.yml \
    ./R-package/configure \
    ./R-package/pkgdown/_pkgdown.yml \
    ./VERSION.txt \
    ./python-package/pyproject.toml

commit_msg="bump development version to ${DEV_VERSION}"
git commit -m "${commit_msg}"
git push upstream ci/dev-version

gh pr create \
    --repo microsoft/LightGBM \
    --base 'master' \
    --label 'maintenance' \
    --title "${commit_msg}" \
    --body "now that v${RELEASE_VERSION} has been released: #${RELEASE_PR_NUMBER}"
```

</details>

### Notes for Reviewers

I believe this should be `v4.5.0` instead of `v4.4.1` because of some small breaking changes:

* #6487
* #6534

See https://github.com/microsoft/LightGBM/releases.

### CODE REVIEW

File: python-package/pyproject.toml
Comment (StrikerRUS) [inline python-package/pyproject.toml:33]: I guess it was a typo.
```suggestion
version = "4.5.0"
```

Comment (jameslamb) [reply python-package/pyproject.toml:33]: ah!! yes, thank you. Maybe in the future we should have a script that automates these changes, like my colleagues do in the RAPIDS repos: https://github.com/rapidsai/cuml/blob/branch-24.08/ci/release/update-version.sh

Comment (StrikerRUS) [reply python-package/pyproject.toml]: Great idea!

**Comment by StrikerRUS** [comment]
> I believe this should be v4.5.0 instead of v4.4.0 because of some small breaking changes:

I guess you meant `v4.4.1`. Agree with your proposal!

**Comment by jameslamb** [comment]
Yes I meant "instead of v4.4.1", just edited it. Sorry for all the typos 😅

**Comment by StrikerRUS** [comment]
Seems that Azure Pipelines builds are constantly failing for the `master` branch.

image

For example, the latest one link: https://dev.azure.com/lightgbm-ci/lightgbm-ci/_build/results?buildId=16597&view=results

Failing jobs are:
- macOS cpp_tests
```
[100%] Built target testlightgbm
/Applications/Xcode_13.1.0.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX12.0.sdk/usr/include/c++/v1/locale:1417:26: runtime error: member call on address 0x612000003ad8 which does not point to an object of type 'std::streambuf'
0x612000003ad8: note: object is of type 'std::__1::basic_stringbuf<char, std::__1::char_traits<char>, std::__1::allocator<char> >'
 01 00 00 00  30 40 cf 0c 01 00 00 00  30 ba 41 5b f8 7f 00 00  00 00 00 00 00 00 00 00  00 00 00 00
              ^~~~~~~~~~~~~~~~~~~~~~~
              vptr for 'std::__1::basic_stringbuf<char, std::__1::char_traits<char>, std::__1::allocator<char> >'
SUMMARY: UndefinedBehaviorSanitizer: undefined-behavior /Applications/Xcode_13.1.0.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX12.0.sdk/usr/include/c++/v1/locale:1417:26 in 
/Users/runner/work/1/s/.ci/test.sh: line 63:  8016 Abort trap: 6           ./testlightgbm

##[error]Bash exited with code '1'.
```
- Windows bdist
```
Installing collected packages: pydistcheck
Successfully installed pydistcheck-0.7.1
Traceback (most recent call last):
  File "<frozen runpy>", line 198, in _run_module_as_main
  File "<frozen runpy>", line 88, in _run_code
  File "C:\Miniconda\envs\test-env\Scripts\pydistcheck.exe\__main__.py", line 7, in <module>
  File "C:\Miniconda\envs\test-env\Lib\site-packages\click\core.py", line 1157, in __call__
    return self.main(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Miniconda\envs\test-env\Lib\site-packages\click\core.py", line 1078, in main
    rv = self.invoke(ctx)
         ^^^^^^^^^^^^^^^^
  File "C:\Miniconda\envs\test-env\Lib\site-packages\click\core.py", line 1434, in invoke
    return ctx.invoke(self.callback, **ctx.params)
```

Should these fails be a blockers for this release?

**Comment by jameslamb** [comment]
> Add release branch to [RTD versions](https://readthedocs.org/projects/lightgbm/versions/), trigger a new build, check docs

* successful build: https://readthedocs.org/projects/lightgbm/builds/24997842/
* docs: https://lightgbm.readthedocs.io/en/release-v4.5.0/

I'll check again once we're done merging changes here.

**Comment by jameslamb** [comment]
> *constantly failing on the `master` branch*

Oh wow, yes I hadn't realized that most `master` builds for the last 2 weeks have failed with exactly those two issues! I just clicked through https://dev.azure.com/lightgbm-ci/lightgbm-ci/_build?definitionId=1&_a=summary&repositoryFilter=1&branchFilter=2%2C2%2C2%2C2%2C2 to see that.

I do think we should try to figure this out for the release, but let's not block releasing the R package on them. If we get to, say, July 30 and have still not fixed it, I think we should at least do a release of the R package to CRAN, to avoid it being archived there.

Could you document these CI failures in a separate issue, and we could discuss further there?

I'm linking this relevant thing I found, that suggests that one issue could be a very long filepath on Windows: https://github.com/sarugaku/shellingham/issues/8. Maybe that could explain the jobs only failing on `master`... maybe something about that changes the filepaths used by Azure DevOps or something? I'm not sure.

**Comment by StrikerRUS** [comment]
> Could you document these CI failures in a separate issue, and we could discuss further there?

Sure, done!
- #6543
- #6544

> If we get to, say, July 30 and have still not fixed it, I think we should at least do a release of the R package to CRAN, to avoid it being archived there.

Absolutely agree!

**Comment by jameslamb** [comment]
I received the following email from CRAN (Prof Brian Ripley) 3 days ago.

> **subject: CRAN packages with g++ 14.x warnings**
>
> _[lightgbm was] given a month's notice with a deadline of Aug 3.  As the CRAN
summer shutdown starts on Aug 1, please submit a corrected update no
later than July 31._

I believe we should release the current state of this PR as v4.5.0 as soon as possible. I want to allow enough time for CRAN to reject the package and for us to put up another submission prior to their summer shutdown.

I do not think there are any other open PRs or issues that absolutely MUST be addressed in this release.

@guolinke @borchero @jmoralez @shiyu1994 @StrikerRUS could you please review, and approve if you agree?

I'd like to get at least 2 approvals (and at least 1 from either @shiyu1994 or @guolinke ) before releasing.

**Review by borchero — APPROVED** [review]
Let's go 🚀

**Review by StrikerRUS — APPROVED** [review]
@jameslamb 
Sure, let's release!

**Comment by jameslamb** [comment]
@guolinke or @shiyu1994 could you please review? I'd like to try to submit to CRAN as soon as possible, and really think one of you 2 should always approve a release.

If you'd like to change that in the future and are ok with me + some other maintainers doing minor releases without requiring your approval, let me know.

**Review by guolinke — APPROVED** [review]

**Comment by StrikerRUS** [comment]
I think we should stop merging any non-trivial PRs before CRAN response. It should help to avoid situation similar to the following one: #4930.

**Comment by jameslamb** [comment]
Agreed!

I'll merge this so we can start the process of releasing. And I'll post follow-up comments here as I've done in previous releases.

Thanks for the reviews and help everyone!

**Comment by jameslamb** [comment]
Ran the following to create the `v4.5.0` tag and update the `stable` tag.

```shell
git checkout master
git pull upstream master
git fetch upstream --tags
git tag -d stable
git push upstream :refs/tags/stable
git tag stable
git tag v4.5.0
git push upstream stable v4.5.0
```

*(NOTE: I alias this repo to `upstream` and my fork to `origin` in my `git` settings)*

https://github.com/microsoft/LightGBM/tags

image

That triggered an Azure DevOps build which should create the release automatically: https://dev.azure.com/lightgbm-ci/lightgbm-ci/_build/results?buildId=16722&view=results

This takes around 90 minutes (because of the QEMU CI job).

I'll check back in a few hours and continue the release then.

**Comment by jameslamb** [comment]
One test job on the release failed, with what looked like a conda network error. Restarted it.

**Comment by jameslamb** [comment]
> Remove the release branch `release/v{n}.{n}.{n}` from RTD versions

done

> Add release notes from release-drafter to the automatically-created GitHub release, convert it from Draft to published.

done!

The release is now live: https://github.com/microsoft/LightGBM/releases/tag/v4.5.0

**Comment by jameslamb** [comment]
Submitted to CRAN with this additional note:

> This release fixes the C++20 issues noted in the "gcc-12" checks at https://cran.r-project.org/web/checks/check_results_lightgbm.html. It is being submitted in response to a request from Prof. Brian Ripley to submit before July 31st.

image

**Comment by StrikerRUS** [comment]
Can't find  `..macosx_..._arm64.whl` macOS wheel for M chips in attached artifacts 🤔 
Do we build them?
Seems that we did it for `4.4.0`:
image

**Comment by StrikerRUS** [comment]
Previously failing CRAN gcc check is now OK! 🎉 

https://cran.r-project.org/web/checks/check_results_lightgbm.html
image

**Comment by jameslamb** [comment]
> Add new tag to [RTD versions](https://readthedocs.org/projects/lightgbm/versions/) and trigger a new build.

This is done.

* build: https://readthedocs.org/projects/lightgbm/builds/25117259/
* docs: https://lightgbm.readthedocs.io/en/v4.5.0/

Screenshot 2024-07-26 at 8 40 10 AM

**Comment by jameslamb** [comment]
> Upload release to [test PyPI](https://test.pypi.org/project/lightgbm/)

This is done: https://test.pypi.org/project/lightgbm/4.5.0/

And installing it (on my M2 Mac) worked fine.

```shell
pip install \
    --no-deps \
    -i https://test.pypi.org/simple/ \
    lightgbm==4.5.0

python -c "import lightgbm"
python examples/python-guide/sklearn_example.py
```

**Comment by StrikerRUS** [comment]
@jameslamb 
>They're built on GitHub Actions because (at least as of https://github.com/microsoft/LightGBM/pull/6391), Azure DevOps didn't have an arm64 runner image.

Ah, I see! Thanks for the explanation!

\* Just checked - still no arm64 macOS runner at Azure 😞

**Comment by jameslamb** [comment]
> Upload release to [PyPI](https://pypi.org/project/lightgbm/).

this is done: https://pypi.org/project/lightgbm/4.5.0/#files

File: python-package/pyproject.toml
Comment (jameslamb) [reply python-package/pyproject.toml]: Very sorry for resolving this without answering you @borchero , that was unintentional.

I'm -1 on using tools like `setuptools-scm` or `versioneer` here. In my experience, they introduce much more complexity than they reduce, at least for projects like LightGBM with a linear sequence of releases (no long-term-support branches, backporting, etc.).

For example, they can be the source of hard-to-debug differences between local and CI builds, unless you're diligent about running `git fetch upstream --tags` or similar regularly.

You might find some of the other discussion in https://discuss.python.org/t/please-make-package-version-go-away/58501 interesting (although I don't agree with the original post... I believe we should continue to support `lightgbm.__version__`).

**Comment by jameslamb** [comment]
>  put up another PR incrementing development version here in GitHub

done: #6575

**Comment by StrikerRUS** [comment]
I think we should add one more release TODO item: "update conda-forge version" since @jameslamb is now official lightgbm feedstock's maintainer.

https://github.com/conda-forge/lightgbm-feedstock/pull/60

**Comment by jameslamb** [comment]
Agreed! I just added it here so we'll remember it when this PR's description is copied into the next release PR.

**Comment by jameslamb** [comment]
> Update [cran-comments.md](https://github.com/microsoft/LightGBM/blob/master/R-package/cran-comments.md) when new release is accepted on CRAN.

put up a PR: https://github.com/microsoft/LightGBM/pull/6578

**Comment by jameslamb** [comment]
> Upload release to [NuGet](https://www.nuget.org/packages/LightGBM/)

done: https://www.nuget.org/packages/LightGBM/4.5.0

That should be fully published some time in the next few hours.

image

**Comment by jameslamb** [comment]
With that, I believe this release is complete! Thanks for the help, everyone.

Really great having you back to help @StrikerRUS , the throughput in this project was great over the last month. This release covers just 6 weeks of changes, but included 49 commits from 16 different contributors (https://github.com/microsoft/LightGBM/compare/v4.4.0...v4.5.0). I'm very happy about that!

**Comment by StrikerRUS** [comment]
@jameslamb Thank you very much for your hard work on the release! 💪

**Pull request merged**
